### DISCUSSION TITLE
How to implement light/dark theme toggle?

### DISCUSSION THREAD

**Discussion by IsaacHub** [NG-ZORRO/ng-zorro-antd#5491] [Q&A] [answered]
## What problem does this feature solve?
The instruction in this page [https://ng.ant.design/docs/customize-theme/en#switch-theming](https://ng.ant.design/docs/customize-theme/en#switch-theming) is not clear, it would be great if you provide a "stackblitz" example.

## What does the proposed API look like?
There are 2 separate pieces of code provided in [https://ng.ant.design/docs/customize-theme/en#switch-theming](https://ng.ant.design/docs/customize-theme/en#switch-theming) , but I have no idea what should be the file name, where should it be placed? How and when should it be called? It will be great if you update that page with an example of doing that on **stackblitz**

**Comment by vthinkxie** [comment]
https://stackoverflow.com/questions/61901234/how-to-change-the-ng-zorro-theme-at-runtime

**Comment by IsaacHub** [comment, marked as answer]
Here is the enterprise-grade professional way of doing it using css variables 
https://stackblitz.com/edit/dark-mode-with-angular-and-css-variables-4bjhu3

**Reply by jdang67** [reply]
Well, we know how to this but we want to see how ng-zorror handle theme

**Reply by KaranTrivedi** [reply]
How is this reply relevant to this forum? This question is in regard to ngZorro library.

**Comment by BeginsDuang** [comment]
```
# yarn
yarn add less -D less-plugin-clean-css -D less-plugin-npm-import -D
# npm
# npm i less -D less-plugin-clean-css -D less-plugin-npm-import -D
```

> project root: create `theme.js`

``` typescript
const less = require("less");
const LessPluginCleanCSS = require("less-plugin-clean-css");
const LessPluginNpmImport = require("less-plugin-npm-import");
const fs = require("fs");
const darkThemeVars = require("ng-zorro-antd/compact-theme");
const compactThemeVars = require("ng-zorro-antd/dark-theme");
const appStyles = "src/assets/styles/zorro.less"; // 应用的样式入口文件 style entry path for the application
const themeContent = `@import '${appStyles}';`;

function gen(type) {
  return less
    .render(themeContent, {
      javascriptEnabled: true,
      plugins: [
        new LessPluginNpmImport({ prefix: "~" }),
        new LessPluginCleanCSS({ advanced: true }),
      ],
      modifyVars: {
        ...(type === "dark" ? darkThemeVars : compactThemeVars),
      },
    })
    .then((data) => {
      fs.writeFileSync(
        // 主题样式的输出文件 output path for the theme style
        `src/assets/themes/style.${type}.css`,
        data.css
      );
    })
    .catch((e) => {
      // 记录渲染错误 log the render error
      console.error(type, e);
    });
}

Promise.all([gen("dark"), gen("compact")]).then(() => {
  console.log("Success!");
});

```

> terminal run：

```
node ./theme.js
```
tips
```
//zorro.less or default src/styles.less
@import "~ng-zorro-antd/ng-zorro-antd.less";  //working
// @import "../../../node_modules/ng-zorro-antd/ng-zorro-antd.css";  //error
```

**Comment by aacebo** [comment]
what I would suggest is creating classes with the nested imports and adding/removing said theme classes from the document body like so:

```
// styles.scss
@import "~ng-zorro-antd/ng-zorro-antd.compact";

body {
  &.ant-dark {
    @import "~ng-zorro-antd/ng-zorro-antd.dark";
  }
}
```

**Comment by Sina7312** [comment]
The easier way could be just to create two theme files e.g. `theme-default.less` and `theme-dark.less`. default theme imports default `ng-zorro` styles and the dark theme imports dark `ng-zorro` styles. include them both in the`styles` section of `angular.json`. You just need to scope the dark styles under some class or something for them to applied conditionally. like this:

`theme-default.less`:
```
@import '~ng-zorro-antd/ng-zorro-antd.less';
// default vars
@primary-color: #3f87f5;
```

`theme-dark.less`:
```
body.dark {
  @import '~ng-zorro-antd/ng-zorro-antd.dark.less';
  // dark vars
  @primary-color: #3f87f5;
}
```

Now you just have to apply the `.dark` class to the `body` to change the theme to dark.

**Reply by zulfiqarlaili** [reply]
I would like to know more about this sir, can you explain a bit more details how to implement this dynamically
